x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -482,6 +487,27 @@ private static List<NamedExpression> combineProjections(
             return replaced;
         }
 
+        private static List<Expression> combineUpperGroupingsAndLowerProjections(
+            List<? extends Expression> upperGroupings,
+            List<? extends NamedExpression> lowerProjections
+        ) {
+            // Collect the alias map for resolving the source (f1 = 1, f2 = f1, etc..)
+            AttributeMap<Expression> aliases = new AttributeMap<>();
+            for (NamedExpression ne : lowerProjections) {
+                // record the alias
+                aliases.put(ne.toAttribute(), Alias.unwrap(ne));
+            }
+
+            // Replace any matching attribute directly with the aliased attribute from the projection.
+            AttributeSet replaced = new AttributeSet();
+            for (Expression expr : upperGroupings) {
+                // All substitutions happen before; groupings must be attributes at this point.
+                Attribute attr = (Attribute) expr;
+                replaced.add((Attribute) aliases.resolve(attr, attr));
```

Comment:
This will lead to a CCE - if there's always an attribute inside the alias (what about c = a + 1), make the aliases be a `AttributeMap<Attribute>`

Reply:
This shouldn't cause CCE in practice, but I can use an `AttributeMap<Attribute>`. (Still need to cast the expressions under the projection's aliases to attributes, then.)

(I'd prefer not having to cast at all, but our Aggs/Projections don't adequately represent the fact that we don't have general expressions here, but attributes.)